batchproducer/batchproducer.go

```diff
@@ -0,0 +1,314 @@
+package batchproducer
+
+import (
+	"errors"
+	"log"
+	"time"
+
+	"github.com/timehop/go-kinesis"
+)
+
+// BatchProducer collects records individually and then sends them to Kinesis in
+// batches in the background using PutRecords, with retries.
+// A BatchProducer will do nothing until Start is called.
+type BatchProducer interface {
+	// Start starts the main goroutine. No need to call it using `go`.
+	Start() error
+	Stop() error
+
+	// Add might block if the BatchProducer has a buffer and the buffer is full.
+	// In order to prevent filling the buffer and eventually blocking indefinitely,
+	// Add will fail and return an error if the BatchProducer is not started.
+	Add(data []byte, partitionKey string) error
+
+	// Setters
+	SetMaxAttemptsPerRecord(int) error
+	SetStatReceiver(BatchStatReceiver) error
+
+	// This interval will be used to make a *best effort* attempt to send stats *approximately*
+	// when this interval elapses. There’s no guarantee, however, since the main goroutine is
+	// used to send the stats and therefore there may be some skew.
+	SetStatInterval(time.Duration) error
+}
+
+// BatchStatReceiver defines an object that can accept stats.
+type BatchStatReceiver interface {
+	// Receive will be called by the main BatchProducer goroutine so it will block all batches from being
+	// sent, so make sure it is either very fast or never blocks at all!
+	Receive(BatchStat)
+}
+
+// BatchStat is a kind of a snapshot of activity and happenings. Some of its fields represent
+// "moment-in-time" values e.g. BufferSize is the size of the buffer at the moment the Stat is
+// sent. Other fields are cumulative since the last Stat, i.e. ErrorsSinceLastStat.
+type BatchStat struct {
+	// Moment-in-time stats
+	BufferSize int
+
+	// Cumulative stats
+	KinesisErrorsSinceLastStat           int
+	RecordsSentSuccessfullySinceLastStat int
+	RecordsDroppedSinceLastStat          int
+}
+
+// BatchingKinesisClient is a subset of KinesisClient to ease mocking.
+type BatchingKinesisClient interface {
+	PutRecords(args *kinesis.RequestArgs) (resp *kinesis.PutRecordsResp, err error)
+}
+
+// NewBatchProducer creates and returns a BatchProducer.
+// The BatchProducer that is returned will flush a batch to Kinesis whenever either
+// the flushInterval occurs (if flushInterval > 0) or the batchSize is reached,
+// whichever happens first. If the number of records in the buffer is equal to
+// or greater than bufferSize then the Add method will block.
+// TODO: this is too many args. Maybe instead have some defaults and add some setter methods for the less critical
+// params
+func NewBatchProducer(
+	client BatchingKinesisClient,
+	streamName string,
+	bufferSize int,
+	flushInterval time.Duration,
+	batchSize int,
+	logger *log.Logger,
+) (BatchProducer, error) {
+	if batchSize < 1 || batchSize > 500 {
+		return nil, errors.New("batchSize must be between 1 and 500 inclusive")
```

Comment:
As far as I can tell, [constants](https://golang.org/ref/spec#Constants) are basically only allowed to be fairly simple scalar values or “complex” constants which are simple statements involving scalar values and operators.

So if I try something like:

```go
const (
	FooError = errors.New("Foo")
)
```

the compiler complains: `const initializer errors.New("Foo") is not a constant`.

That seems to be the reason that the error values in `io` are variables rather than constants.

So it would seem that if we want to create package-level error values so they can be compared, we’d have to make them variables, unless I’m missing something.

Personally this would seem to me at this point like premature optimization and/or a violation of YAGNI… but anyway, would you still like me to proceed?